### PR TITLE
NEAR Simulator -> workspaces-rs migration guide

### PR DESCRIPTION
**Pull request by itegulov**
I was debating whether to go the simple route or try [multi-language blocks](https://docusaurus.io/docs/markdown-features/code-blocks#multi-language-support-code-blocks) to create a direct comparison between near-sdk-sim/workspaces-rs code, but ended with the former. Let me know if you would be interested in seeing how the alternative looks like.

Feel free to point out **any** (however minor or big) spelling/grammar/stylistic errors you can see. English is not my native language and my writing style is pretty stiff, so all input is appreciated. Any other feedback is of course welcome as well!

### CODE REVIEW

File: docs/testing/workspaces-migration-guide.md [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+sidebar_label: Workspaces Migration Guide
+title: "Workspaces Migration Guide"
+---
+
+# Workspaces Migration Guide
+
+NEAR Simulator was meant to be an in-place replacement of a blockchain environment for the purpose of testing NEAR contracts. However, simulating NEAR ledger turned out to be a much more complex endeavour than was anticipated. Eventually, the idea of workspaces was born - a library for automating workflows and writing tests for NEAR smart contracts using a real NEAR network (localnet, testnet or mainnet). Thus, NEAR Simulator is being deprecated in favor of [`workspaces-rs`](https://github.com/near/workspaces-rs), the Rust edition of workspaces. As the two libraries have two vastly different APIs this guide was created to ease the migration process for developers.
+
+:::danger
+TODO: I do not have a whole lot of context here why exactly simtests were not suitable for our purposes, so if anyone wants to elaborate the preceding paragraph please do.
```
Comment (itegulov) [inline docs/testing/workspaces-migration-guide.md:12]: Can someone help filling this gap? @ChaoticTempest you probably have the most amount of context here, feel free to write it as a comment or push a commit to my branch.

File: docs/testing/workspaces-migration-guide.md [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    .root_account()
+    .create_subaccount(&worker, id)
+    .initial_balance(parse_near!("5 N"))
+    .transact()
+    .await?
+    .result;
+```
+
+:::danger
+TODO: Is there a reason why we can't control the initial balance with `dev_deploy`?
```
Comment (itegulov) [inline docs/testing/workspaces-migration-guide.md:119]: @ChaoticTempest question to you

File: docs/testing/workspaces-migration-guide.md [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+```rust title="Logs - workspaces-rs"
+let outcome = &res.receipt_outcomes()[5];
+assert_eq!(outcome.logs[0], "The account of the sender was deleted");
+assert_eq!(outcome.logs[2], format!("Account @{} burned {}", contract.id(), 10));
+```
+
+## Profiling gas
+
+NEAR Simulator has not been providing realistic gas estimates (TODO: needs elaboration - since when and why?), so the `workspaces-rs` gas estimates might come as a surprise. The good news is that they do reflect reality!
```
Comment (itegulov) [inline docs/testing/workspaces-migration-guide.md:300]: @ChaoticTempest also probably yours :)

File: docs/testing/workspaces-migration-guide.md (hunk 1 above)
Comment (austinabell) [reply docs/testing/workspaces-migration-guide.md:12]: A few reasons I can remember:
- near-sdk-sim was hooking into parts of nearcore that were not meant to be released, in the most recent version those crates aren't released so sim is currently using duplicate deps (maintenance nightmare)
- Not a fully accurate simulation because it just used a subset of the runtime in a specific way -- can't rely on this
- Bulky dependencies pulled in (drastically increases compile time)
- Unergonomic API, not specific to this strategy, but likely would have had to be re-built
- Can't test parallel transactions easily -- current pattern would process blocks until a transaction succeeded but you can't create specific conditions, which is required for a strategy like this that isn't fully simulated

I'll let Phuong share opinions when he's back since he likely will have some inputs for this

Comment (ChaoticTempest) [reply docs/testing/workspaces-migration-guide.md:12]: > Not a fully accurate simulation because it just used a subset of the runtime in a specific way -- can't rely on this

And thus couldn't measure gas burnt accurately. Also, all the intricacies of nearcore (like protocol features) wouldn't be one-to-one with the runtime since the runtime was just code built on top of VM logic. 

People would also need to write their own automation scripts to deploy to testnet, so we'd end up with very split workflows for testing.

File: docs/testing/workspaces-migration-guide.md (hunk 2 above)
Comment (ChaoticTempest) [reply docs/testing/workspaces-migration-guide.md:119]: `dev_deploy` can't supply the initial balance since testnet controls this amount in the helper contract which is what we're using to create dev accounts in testnet. So, to make it simple, we don't supply it at all for sandbox even. If someone wants to supply an amount in sandbox, they can grab the root account and do:
```rust
let root = worker.root_acount();
root.create_subaccount(...)
   .initial_balance(...)
   ...
```

File: docs/testing/workspaces-migration-guide.md (hunk 3 above)
Comment (ChaoticTempest) [reply docs/testing/workspaces-migration-guide.md:300]: > TODO: needs elaboration - since when and why?

it never had accurate gas estimations since it only tried to mirror nearcore, but nearcore has extra things on top which consumes gas (like cross-contract calls are processed separately from the same transaction and that incurs gas fees)

I'd give a warning that since workspaces-rs is now using accurate gas measurements, some contracts that were previously being tested with sdk-sim might not work the same, and that they should do their due diligence if they plan to deploy to mainnet. I mean, either ways, if they tried deploying the same contracts with infeasible gas estimates to testnet, it would have never worked. So we can say that workspaces offers the better experience here and aligns very well with what you can do on testnet and mainnet

File: docs/testing/workspaces-migration-guide.md [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+
+const ID: &str = "contract-id";
+
+...
+
+let root = init_simulator(...);
+let contract = root.deploy(&WASM_BYTES, ID.parse().unwrap(), to_yocto("5"));
+```
+
+Although `workspaces-rs` provides a way to specify the account id for a contract to be deployed, usually it does not matter in the context of a single test. If you are fine with generating a random developer account and initializing it with 100N, then you can use replace the snippet above with this:
```
Comment (ChaoticTempest) [inline docs/testing/workspaces-migration-guide.md:79]:
```suggestion
Although `workspaces-rs` provides a way to specify the account id for a contract to be deployed, usually it does not matter in the context of a single test. If you are fine with generating a random developer account and initializing it with 100N, then you can replace the snippet above with this:
```

File: docs/testing/workspaces-migration-guide.md [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+let root_balance: U128 = contract
+    .call(&worker, "ft_balance_of")
+    .args_json((contract.id(),))?
+    .view()
+    .await?
+    .json()?;
+```
+
+:::note
+Note that you have to pass arguments as any serializable type representing a sequential list. Tuples are usually the best candidate due to their heterogeneous nature (remember that you can construct a unary tuple by placing a comma before the closing bracket like this: `(el,)`).
```
Comment (ChaoticTempest) [inline docs/testing/workspaces-migration-guide.md:196]: Might want to also mention the `json!()` macro is acceptable here as well

File: docs/testing/workspaces-migration-guide.md [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+            .deposit(1),
+    )
+    .transact()
+    .await?;
+```
+
+## Inspecting logs
+
+:::danger
+TODO: I found some potential inconsistencies in what I wrote about the log inspection behavior, need some time to look at this further
```
Comment (ChaoticTempest) [inline docs/testing/workspaces-migration-guide.md:255]: what kind of inconsistencies did you find. The section you wrote afterwards describes it well

File: docs/testing/workspaces-migration-guide.md [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+let worker = workspaces::sandbox().await?;
+let id: AccountId = "contract-id".parse()?;
+let contract = worker
+    .root_account()
+    .create_subaccount(&worker, id)
+    .initial_balance(parse_near!("5 N"))
+    .transact()
+    .await?
+    .result;
```
Comment (ChaoticTempest) [inline docs/testing/workspaces-migration-guide.md:115]: This doesn't actually deploy the contract, since this only creates the subaccount. Also, better to use `into_result()?` instead of `.result`

File: docs/testing/workspaces-migration-guide.md [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+:::danger
+TODO: I do not have a whole lot of context here why exactly simtests were not suitable for our purposes, so if anyone wants to elaborate the preceding paragraph please do.
+:::
+
+This guide presumes that you are transitioning from near-sdk-sim `3.2.0` (the last non-deprecated release) to `workspaces-rs` `0.2.1`. Given that near-sdk-sim is deprecated, it is very unlikely that its API is going to ever change, but future releases of `workspaces-rs` might. Hopefully this guide is going to be helpful even if you are migrating your project to a more recent version, but also feel free to migrate your tests to `0.2.1` using this guide first and upgrade to the most recent workspaces-rs version later by looking at the release notes to see how public API has changed since `0.2.1`.
```
Comment (ChaoticTempest) [inline docs/testing/workspaces-migration-guide.md:15]: nit: this paragraph is a bit wordy. Maybe something like this would be better

```suggestion
This guide presumes that you are transitioning from near-sdk-sim `3.2.0` (the last non-deprecated release) to `workspaces-rs` `0.2.1`. Given that near-sdk-sim is deprecated, it is very unlikely that its API will ever change, but future releases of `workspaces-rs` might. Hopefully, this guide will be helpful even if you are migrating your project to a more recent workspaces version. If workspaces have changed, feel free to migrate your tests to `0.2.1` first using this guide and upgrade to the most recent workspaces-rs version later by looking at the release notes to see how public API has changed since `0.2.1`.
```

**Review by ChaoticTempest — COMMENTED** [review]
Just a couple nits, but overall looking pretty swell!

cc @AustinBaggio who might want to take a look or rope in the DevRel guys with this one

**Comment by AustinBaggio** [comment]
Hey @doriancrutcher do we have an existing sim-test project we can test this migration plan on? cc @thisisjoshford

File: docs/testing/workspaces-migration-guide.md (hunk 8 above)
Comment (itegulov) [reply docs/testing/workspaces-migration-guide.md]: Yeah, that is much nicer

**Comment by idea404** [comment]
I've implemented the content and feedback from this PR in another branch I just merged with this one. 

What do you guys think about it now? @BenKurrek @itegulov @ChaoticTempest @austinabell

File: docs/testing/integration-tests.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+```sh
+├── Cargo.toml                  ⟵ contains `dependencies` for contract and `dev-dependencies` for workspaces-rs tests
+├── src
+│  └── lib.rs                   ⟵ contract code
+├── target
+└── tests                       ⟵ integration test directory
+   └── integration-tests.rs     ⟵ integration test file
+```
+
+A sample configuration for theis project's `Cargo.toml` is shown below:
```
Comment (ChaoticTempest) [inline docs/testing/integration-tests.md:38]:
```suggestion
A sample configuration for this project's `Cargo.toml` is shown below:
```

**Review by ChaoticTempest — APPROVED** [review]
LGTM besides one typo

File: docs/testing/integration-tests.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+- There are cross-contract calls.
+- There are multiple users with balance changes.
+- You'd like to gather information about gas usage and execution outcomes on-chain.
+- You want to assert the use-case execution flow of your smart contract logic works as expected.
+- You want to assert given execution patterns do not work (as expected).
+
+## Setup
+
+Unlike unit tests (which would often live in the `src/lib.rs` file of the contract), integration tests in Rust are located in a separate directory at the same level as `/src`, called `/tests` ([read more](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#integration-tests)). Refer to this folder structure below:
```
Comment (austinabell) [inline docs/testing/integration-tests.md:27]: They don't have to be. Can have workspaces tests within the src directory too. This can be beneficial because then you can use the non-exported types for serialization within the test case.

File: docs/testing/integration-tests.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+:::note
+In case you wish to preserve state between runs, you can call multiple tests within one function, passing the worker around from a `workspaces::sandbox()` call.
+:::
+
+## Helpful Snippets
+
+### Create an Account
+
+```rust reference
+https://github.com/near-examples/rust-counter/blob/6a7af5a32c630e0298c09c24eab87267746552b2/integration-tests/rs/src/tests.rs#L16-L21
```
Comment (austinabell) [inline docs/testing/integration-tests.md:122]: maybe in future also indicating the ability to dev_create_account? cc @ChaoticTempest do you think this is worthwhile?

**Review by austinabell — APPROVED** [review]
This seems fine for now. Probably best we get feedback from new users on this, I don't have a high confidence this will be intuitive or easy to follow, especially with the raw amount of code that is included

File: docs/testing/integration-tests.md (hunk 10 above)
Comment (idea404) [reply docs/testing/integration-tests.md:27]: I added an `info` clause below mentioning this, thank you!

File: docs/testing/integration-tests.md (hunk 11 above)
Comment (idea404) [reply docs/testing/integration-tests.md:122]: I included a note on `dev_create_account` in this section, thanks guys!
